python/cudf/cudf/core/algorithms.py

```diff
@@ -7,47 +7,111 @@
+        If False, NA values will be encoded as non-negative
+        integers and will not drop the NA from the uniques
+        of the values.
+
     Returns
     -------
     (labels, cats) : (cupy.ndarray, cupy.ndarray or Index)
         - *labels* contains the encoded values
         - *cats* contains the categories in order that the N-th
             item corresponds to the (N-1) code.
 
+    See Also
+    --------
+    cudf.Series.factorize : Encode the input values of Series.
+
     Examples
     --------
     >>> import cudf
+    >>> import numpy as np
     >>> data = cudf.Series(['a', 'c', 'c'])
     >>> codes, uniques = cudf.factorize(data)
     >>> codes
     array([0, 1, 1], dtype=int8)
     >>> uniques
     StringIndex(['a' 'c'], dtype='object')
 
-    See Also
-    --------
-    cudf.Series.factorize : Encode the input values of Series.
+    When ``use_na_sentinel=True`` (the default), missing values are indicated
+    in the `codes` with the sentinel value ``-1`` and missing values are not
+    included in `uniques`.
+
+    >>> codes, uniques = cudf.factorize(['b', None, 'a', 'c', 'b'])
+    >>> codes
+    array([ 1, -1,  0,  2,  1], dtype=int8)
+    >>> uniques
+    StringIndex(['a' 'b' 'c'], dtype='object')
 
+    If NA is in the values, and we want to include NA in the uniques of the
+    values, it can be achieved by setting ``use_na_sentinel=False``.
+
+    >>> values = np.array([1, 2, 1, np.nan])
+    >>> codes, uniques = cudf.factorize(values)
+    >>> codes
+    array([ 0,  1,  0, -1], dtype=int8)
+    >>> uniques
+    Float64Index([1.0, 2.0], dtype='float64')
+    >>> codes, uniques = cudf.factorize(values, use_na_sentinel=False)
+    >>> codes
+    array([1, 2, 1, 0], dtype=int8)
+    >>> uniques
+    Float64Index([<NA>, 1.0, 2.0], dtype='float64')
     """
-    if sort:
-        raise NotImplementedError(
-            "Sorting not yet supported during factorization."
+    # TODO: Drop `na_sentinel` in the next release immediately after
+    # pandas 2.0 upgrade.
+    if na_sentinel is not None:
+        warnings.warn(
+            "Specifying the specific value to use for `na_sentinel` is "
+            "deprecated and will be removed in a future version of cudf. "
+            "Specify `use_na_sentinel=True` to use the sentinel value -1, "
+            "and `use_na_sentinel=False` to encode NA values.",
+            FutureWarning,
+        )
+
+    if use_na_sentinel is not None and na_sentinel is not None:
+        raise ValueError(
+            "Cannot specify both `na_sentinel` and `use_na_sentile`; "
+            f"got `na_sentinel={na_sentinel}` and "
+            f"`use_na_sentinel={use_na_sentinel}`"
+        )
+    elif use_na_sentinel is None and na_sentinel is None:
+        use_na_sentinel = True
+        na_sentinel = -1
+
+    if use_na_sentinel is None:
+        use_na_sentinel = True
+    elif na_sentinel is None:
+        na_sentinel = -1
+    else:
```

Comment:
The logic in this function is a bit difficult to follow (at least for me), with all the `if`s and `else`s.

If I'm reading correctly, the `else` is reached when:

```suggestion
    else:  # use_sentinel is either True or False, na_sentinel is None
```

Is that correct? if so, I think the comment helps clarify.

Reply:
Yea, it's a lot of code smell here. Let me give another try to simplify things here today.

Reply:
Simplified the logic, let me know what you think about it now.

Reply:
FWIW pandas had some mangled code when deprecating this in 1.5 https://github.com/pandas-dev/pandas/pull/47157/files#diff-c8f3ad29eaf121537b999e88e9117f3e3702d0b818a67516da25093fe2890ce8R794